java-checks/src/main/java/org/sonar/java/checks/aws/AwsLambdaSyncCallCheck.java

```diff
@@ -0,0 +1,191 @@
+
+    private static final MethodMatchers INVOCATIONTYPE_MATCHERS = MethodMatchers.create()
+      .ofTypes("com.amazonaws.services.lambda.model.InvokeRequest")
+      .names("setInvocationType", "withInvocationType")
+      .addParametersMatcher("java.lang.String").build();
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      getSyncCalls(tree).ifPresent(msgPart -> invokeInvocations.put(tree, msgPart));
+    }
+
+    public Map<MethodInvocationTree, String> getSyncInvokeCalls() {
+      return invokeInvocations;
+    }
+
+    private static Optional<String> getSyncCalls(MethodInvocationTree tree) {
+      if (INVOKE_MATCHERS.matches(tree)) {
+
+        // INVOKE_MATCHER ensures that there is one argument.
+        ExpressionTree argument = tree.arguments().get(0);
+
+        if (!argument.is(Tree.Kind.IDENTIFIER)) {
+          return Optional.empty();
+        }
+        IdentifierTree invokeRequest = (IdentifierTree) argument;
+
+        // We know there is at least one usage, i.e. the one we just got above.
+        List<IdentifierTree> localUsages = invokeRequest.symbol().usages().stream()
+          .filter(u -> isLocalVariable(u.symbol()) && !u.equals(invokeRequest))
+          .collect(Collectors.toList());
+
+        if (isParameter(invokeRequest.symbol())
+          || localUsages.stream().anyMatch(lu -> isArgumentToACall(lu) || statementSetsAsyncCall(lu))
+          || declarationSetsAsyncCall(invokeRequest)) {
+          return Optional.empty();
+        }
+
+        if (hasLocalVarDeclaration(invokeRequest)) {
+          return Optional.of(MESSAGE);
+        }
+        return Optional.empty();
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    private static boolean isArgumentToACall(IdentifierTree invokeRequest) {
+      return invokeRequest.parent() != null && invokeRequest.parent().is(Tree.Kind.ARGUMENTS);
+    }
+
+    private static boolean hasLocalVarDeclaration(IdentifierTree invokeRequest) {
+      Tree declaration = invokeRequest.symbol().declaration();
+      return (declaration != null && declaration.is(Tree.Kind.VARIABLE) && isLocalVariable(((VariableTree) declaration).symbol()));
+    }
+
+    /**
+     * Returns true if the statement starting at 'tree' sets the InvocationType object
+     * to a configuration that leads to async lambda calls. The statement could be a sequence of calls.
+     *
+     * @param tree the statement starting point
+     * @return true if statement leads lambda calls to be async
+     */
+    private static boolean statementSetsAsyncCall(Tree tree) {
+      MethodInvocationTree methodCall = findMethodInvocationTreeAncestor(tree);
+      if (methodCall != null) {
+        if (setsInvocationTypeToAsync(methodCall)) {
+          return true;
+        } else {
+          return statementSetsAsyncCall(methodCall);
+        }
+      }
+      return false;
```

Comment:
I believe this could be simplified to:
```suggestion
      return (methodCall != null && (setsInvocationTypeToAsync(methodCall) || statementSetsAsyncCall(methodCall)));
```